src/lua/uri.lua

```diff
@@ -277,11 +283,19 @@ local function str_escape(buf, opts)
     if opts.plus == true then
         url = url:gsub(" ", "+")
     end
-    return url
-end
 
-local hex_to_char = function(x)
-    return string.char(tonumber(x, 16))
+    if #opts.unreserved > 255 then
+        error("str_escape: opts.unreserved should be less that 255")
+    end
+
+    local c_unreserved = ffi.new("char[?]", #opts.unreserved + 1)
+    ffi.copy(c_unreserved, opts.unreserved, #opts.unreserved)
+    local c_str = ffi.new("char[?]", #buf + 1)
+    ffi.copy(c_str, buf, #buf)
+    local c_url_enc = builtin.str_escape(c_str, c_unreserved, opts.plus and 1 or 0)
+    local url = ffi.string(c_url_enc)
```

Comment:
The pure Lua algorithm is kept above: so we encode the string twice using two algorithms and just discard the first result. Maybe LuaJIT eliminates the dead code, but it anyway looks confusing.